core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala

```diff
@@ -144,11 +144,24 @@ private[spark] class PythonRDD(
                 stream.readFully(update)
                 accumulator += Collections.singletonList(update)
               }
+
               // Check whether the worker is ready to be re-used.
-              if (stream.readInt() == SpecialLengths.END_OF_STREAM) {
-                if (reuse_worker) {
-                  env.releasePythonWorker(pythonExec, envVars.toMap, worker)
-                  released = true
+              if (reuse_worker) {
+                // It has a high possibility that the ending mark is already available,
+                // And current task should not be blocked by checking it
+
+                if (stream.available() >= 4) {
+                  val ending = stream.readInt()
+                  if (ending == SpecialLengths.END_OF_STREAM) {
+                    env.releasePythonWorker(pythonExec, envVars.toMap, worker)
+                    released = true
+                    logInfo(s"Socket is ended cleanly, reuse it: $worker")
+                  } else {
+                    logInfo(s"Socket is not ended cleanly (ending with $ending), " +
```

Comment:
The "ended cleanly" case can probably stay at `logInfo` (or maybe `logDebug`), but I think we should make this case and the other error-case into warnings so that they aren't swallowed at lower log levels.